h2/src/main/org/h2/command/Parser.java

```diff
@@ -2773,10 +2775,23 @@ private void parseEndOfQuery(Query command) {
                     do {
                         readIdentifierWithSchema();
                     } while (readIf(COMMA));
-                } else if (readIf("NOWAIT")) {
-                    // TODO parser: select for update nowait: should not wait
                 }
-                command.setForUpdate(true);
+                ForUpdate forUpdate;
+                if (readIf("NOWAIT")) {
+                    forUpdate = ForUpdate.NOWAIT;
+                } else if (readIf("WAIT")) {
+                    BigDecimal timeout;
+                    if (currentTokenType != LITERAL || (timeout = token.value(session).getBigDecimal()) == null) {
+                        throw DbException.getSyntaxError(sqlCommand, token.start(), "numeric");
+                    }
+                    read();
+                    forUpdate = ForUpdate.wait(timeout.multiply(BigDecimal.valueOf(1_000L)).intValue());
```

Comment:
maybe check that timeout is < Integer.MAX_VALUE here, to protect against idiots specifying huge values